### PR TITLE
Fix double optimizer step in DefaultTrainingLoop with static capture

### LINKED ISSUE #1911: 🐛[BUG]: DefaultTrainingLoop performs two optimizer steps per minibatch with static capture
### Version

main branch

### On which installation method(s) does this occur?

Source

## Summary

`DefaultTrainingLoop(enable_static_capture=True)` performs **two optimizer updates for every minibatch**.

The static-capture wrapper owns one update through `GradScaler.step(optimizer)`, then `DefaultTrainingLoop` unconditionally calls `optimizer.step()` again after the wrapper returns. Because the gradients remain attached, the second call advances the parameters and optimizer state again.

The loop still reports one minibatch and advances the scheduler only once.

Verified on `main` at commit `043f63bc70dd3d9a491d611b5e27d41af10647b6`.

## Current code path

1. The active-learning loop wraps the training step in `StaticCaptureTraining(model=model, optim=optimizer, ...)` when static capture is enabled:

   https://github.com/NVIDIA/physicsnemo/blob/043f63bc70dd3d9a491d611b5e27d41af10647b6/physicsnemo/active_learning/loop.py#L368-L379

2. The wrapper backpropagates and calls `self.scaler.step(self.optim)`:

   https://github.com/NVIDIA/physicsnemo/blob/043f63bc70dd3d9a491d611b5e27d41af10647b6/physicsnemo/utils/capture.py#L157-L174

3. After the wrapped call returns, `DefaultTrainingLoop` skips only `loss.backward()` but still calls `optimizer.step()` unconditionally:

   https://github.com/NVIDIA/physicsnemo/blob/043f63bc70dd3d9a491d611b5e27d41af10647b6/physicsnemo/active_learning/loop.py#L509-L521

## Minimal reproduction

This reproduces on CPU; a GPU or active CUDA graph is not required to expose the duplicate optimizer-step ownership.

```python
import torch
from torch.utils.data import DataLoader, TensorDataset

from physicsnemo.active_learning import DefaultTrainingLoop
from physicsnemo.models.mlp import FullyConnected


def run(capture: bool):
    torch.manual_seed(0)
    model = FullyConnected(
        in_features=1,
        out_features=1,
        layer_size=1,
        num_layers=0,
    )

    with torch.no_grad():
        for parameter in model.parameters():
            parameter.fill_(1.0)

    optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
    loader = DataLoader(
        TensorDataset(torch.ones(1, 1), torch.zeros(1, 1)),
        batch_size=1,
    )

    def step(module, batch):
        x, y = batch
        return (module(x) - y).square().mean()

    before = [parameter.detach().clone() for parameter in model.parameters()]

    DefaultTrainingLoop(
        enable_static_capture=capture,
        use_progress_bars=False,
    )(
        model,
        optimizer,
        loader,
        max_epochs=1,
        train_step_fn=step,
        device=torch.device("cpu"),
        dtype=torch.float32,
    )

    after = [parameter.detach().clone() for parameter in model.parameters()]
    return [(new - old).item() for old, new in zip(before, after)]


print("eager:", run(False))
print("capture:", run(True))
```

Observed:

```text
eager:   [-0.4, -0.4]
capture: [-0.8, -0.8]
```

## Expected behavior

Static capture should preserve the eager training semantics: exactly one optimizer update per minibatch.

For this deterministic example, the eager and captured parameter deltas should match.

## Impact

Every use of `DefaultTrainingLoop(enable_static_capture=True)` silently applies an extra optimizer update per minibatch. This:

- changes the effective optimization trajectory;
- advances stateful optimizers twice;
- makes scheduler steps, minibatch counts, logs, and training budgets disagree with the actual update count; and
- can produce apparently valid training results using unintended optimization semantics.

## Suggested fix

Give optimizer-step ownership to exactly one layer. Either:

- suppress the outer `optimizer.step()` when `StaticCaptureTraining` owns backward and step; or
- restructure the capture wrapper so the outer training loop remains the sole optimizer-step owner.

A regression test should run one deterministic minibatch with static capture disabled and enabled, then assert:

1. identical parameter deltas; and
2. exactly one optimizer-step call in both paths.

### Minimum reproducible example

```shell

```

### Relevant log output

```shell

```

### Environment details

```shell

```

### PR DESCRIPTION
**Pull request by wilyan09007**
# PhysicsNeMo Pull Request

## Description

`StaticCaptureTraining` runs the backward pass and steps the optimizer through its grad scaler, but the loop skipped only `loss.backward()` and called `optimizer.step()` unconditionally. The gradients are still populated there, so with `enable_static_capture=True` every minibatch got a second update: the effective learning rate doubled, stateful optimizers advanced twice, and the update bypassed the grad scaler's inf/nan check, while the minibatch count and scheduler still recorded one.

This moves `optimizer.step()` into the eager branch, leaving the capture wrapper the sole owner of the update.

`test_single_optimizer_step_per_minibatch` runs one deterministic minibatch each way on CPU and asserts identical parameter deltas and a single `optimizer.step` call; it fails on main. Every other test in that file sets `enable_static_capture=False`, so this path had no coverage.

Unrelated and left alone: the loop's `model.zero_grad(set_to_none=True)` before each wrapped step is what `capture.py` warns can invalidate a recorded CUDA graph.

Closes #1911

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies

None.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI's assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: test/active_learning/test_loop.py [hunk 1]
```diff
@@ -446,3 +447,51 @@ def call_train_step(model, batch, *args, **kwargs):
...
+
+        def run(enable_static_capture: bool) -> tuple[list[torch.Tensor], int]:
+            mock_module.load_state_dict(initial_state)
+            optimizer = torch.optim.SGD(mock_module.parameters(), lr=0.1)
+            optimizer.step = MagicMock(side_effect=optimizer.step)
+
+            DefaultTrainingLoop(
+                enable_static_capture=enable_static_capture,
+                use_progress_bars=False,
+                use_amp=False,
```
[the remark this answers is not in the document]
Comment (wilyan09007) [reply test/active_learning/test_loop.py:474]: On CPU there is no graph or enabled scaler for this test to skip. `_StaticCapture.__init__` takes the CPU branch, which sets `cuda_graphs_enabled = False` and builds the scaler with `_init_amp_scaler(False)` no matter what `use_amp` is; `use_gradscaler` is only read on the CUDA branch. So turning `use_amp` back on would not reach graph record/replay or an enabled scaler, it would only add CPU bfloat16 autocast, and that is what I turned off so the two paths can be compared with an exact `torch.equal` on the parameter deltas.

Step ownership is covered. A disabled `GradScaler.step(optim)` still calls through to `optimizer.step`, so `call_count == 1` on the capture path is counting the wrapper's own update, which is the behavior the fix depends on.

Enabled-scaler inf/nan skipping and graph replay are GPU-only and belong to `capture.py`, where `test/utils/test_capture.py` already parametrizes over `cuda:0`. What changed here is a branch in the training loop that does not vary by device or capture mode.
